File: C++/MergeSort.cpp
#include <iostream>
#include <vector>
using namespace std;

void merge(vector<int>& A, int p, int q, int r) {
    vector<int> L(A.begin() + p, A.begin() + q + 1);
    vector<int> R(A.begin() + q + 1, A.begin() + r + 1);
    L.push_back(INT32_MAX);
    R.push_back(INT32_MAX);
    int i = 0, j = 0;
    for (int k = p; k <= r; k++) {
        if (L[i] <= R[j]) A[k] = L[i++];
        else A[k] = R[j++];
    }
}

void mergeSort(vector<int>& A, int p, int r) {
    if (p < r) {
        int q = (p + r) / 2;
        mergeSort(A, p, q);
        mergeSort(A, q + 1, r);
        merge(A, p, q, r);
    }
}

int main() {
    vector<int> A = {5, 2, 4, 7, 1, 3, 2, 6};
    mergeSort(A, 0, A.size() - 1);
    for (int x : A) cout << x << " ";
    cout << endl;
    return 0;
}

File: MergeSort.py
import random

def merge_sort(A):
    def merge_sort_recursive(A, p, r):
        if p < r:
            q = (p + r) // 2
            merge_sort_recursive(A, p, q)
            merge_sort_recursive(A, q + 1, r)
            merge(A, p, q, r)

    def merge(A, p, q, r):
        L = A[p:q+1]
        R = A[q+1:r+1]
        i = j = 0
        k = p

        while i < len(L) and j < len(R):
            if L[i] <= R[j]:
                A[k] = L[i]
                i += 1
            else:
                A[k] = R[j]
                j += 1
            k += 1

        while i < len(L):
            A[k] = L[i]
            i += 1
            k += 1
        while j < len(R):
            A[k] = R[j]
            j += 1
            k += 1

    merge_sort_recursive(A, 0, len(A) - 1)

if __name__ == "__main__":
    print("Now displaying MergeSort")
    A = [random.randint(0, 1000) for _ in range(random.randint(5, 100))]
    print("Before:", A)
    merge_sort(A)
    print("After:", A)